enhancements/automatic-catalog-switching-annotations.md

```diff
@@ -0,0 +1,343 @@
+
+These variables allow the user to identify arbitrary values to use for template substitution by pinpointing a 
+specific manifest by GVK, with a given name in a given namespace (if namespace scoped), and pull out any value within that manifest. 
+All of these parameters MUST be provided, and the operator must have permissions to access the resource for 
+this lookup to succeed. 
+
+The GVK with JSON path variables MUST be provided in the exact order shown in the following example, where the 
+variable name is followed by a colon and separated from the next variable by a comma:
+
+```
+{group:foo.example.com,version:v1,kind:Sample,name:MySample,namespace:ns,jsonpath:spec.foo.bar}
+```
+
+Failure to follow this syntax will result in a invalid template which will be ignored. 
+
+Note that a GVK with JSON path template must be surrounded by `{` and `}`, contain no spaces, and follow the case sensitivity rules 
+defined for each parameter above.
+
+### Required well known templates
+
+These templates are well known and can be resolved by any kubernetes distribution.
+
+- {kube_major_version}
+  - example resolved value: 1
+- {kube_minor_version}
+  - example resolved value: 19
+- {kube_patch_version}
+  - example resolved value: 0
+
+### Optional GVK templates with JSON Path
+
+- `{group:foo.example.com,version:v1,kind:Sample,name:MySample,namespace:ns,jsonpath:spec.foo.bar}`
+
+### Templates that have been discussed and rejected
+
+#### OLM version
+
+The OLM version is not consistent between upstream and downstream and therefore unreliable
+
+- {olm_major_version}
+  - example resolved value: 0
+- {olm_minor_version}
+  - example resolved value: 18
+- {olm_patch_version}
+  - example resolved value: 1
+
+#### Platform architecture
+
+Architecture may not be consistent if running in a heterogeneous cluster environment
+
+- {platform_architecture}
+  - example resolved value: x86_64
+
+#### Distribution specific
+
+Declaring distribution specific variables does not scale well
+
+- {ocp_major_version}
+  - example resolved value: 1
+- {ocp_minor_version}
+  - example resolved value: 1
+- {ocp_patch_version}
+  - example resolved value: 1
+- {ocp_update_channels}
+
+#### Distribution version
+
+Declaring generic distribution variables leaves gaps in support since there
+are many distributions to consider and could lead to confusion when a 
+particular distribution is not supported.
+
+- {distro_major_version}
+  - example resolved value: 1
+- {distro_minor_version}
+  - example resolved value: 1
+- {distro_patch_version}
+  - example resolved value: 1
+
+### CatalogSource updates via annotations
+
+The mechanism for declaring an image using templates involves an annotation called
+`olm.catalogImageTemplate` whose value consists of the image reference with one or more templates included.
+A controller (which can exist in either the OLM catalog operator or within a standalone operator) is 
```

Comment:
I am struggling to understand the interaction of the 2 controller approach here, when I initially apply a catalogsource yaml file to the cluster, what happens?  I assume the existing catalog operator controller "sees" the new resource and starts to spin up the associated pod using the original image in the spec, the new controller also sees the new resource and modifies the same resource with new image?  Is there a coordination problem here?

I am also trying to understand the switching part, if we are looking at checking the kube version, when is this actually checked?  Is it when the operator actually restarts and sees a new kube version or is there some sort of polling/cron job kinda thing in play here?  I guess if we are updating the kube version of the cluster it is assumed that operator gets restarted during the upgrade?

Reply:
Just re-read Evans comment from the [other PR](https://github.com/operator-framework/enhancements/pull/86#discussion_r663040871), he did mention polling, so in addition to watching from catalogsources, there is another thread that does the polling and if it finds changes, it then pulls in all the existing catalogsources and applies the new polling results if different?

Also, on the topic of where this controller lives, my vote would be as part of the OLM catalog operator, may be harder to integrate but if its in OLM then customers don't have to do take addition actions to install it

Reply:
> I guess if we are updating the kube version of the cluster it is assumed that operator gets restarted during the upgrade?

In general, I don't think this is a safe assumption to make. OLM itself cannot make any guarantees that some other external system will upgrade it when the cluster apiserver upgrades.

It seems like we'll need to poll the apiserver for its version information to handle the case where the catalog operator pod survives a cluster upgrade.

One possible non-polling solution would be to trigger a new version info request whenever the operator's connection(s) to the apiserver is disconnected. Not totally sure this would work though (e.g. is it possible for there to be a proxy between the operator and the apiserver pod that might keep a connection open even across apiserver restarts?)

Reply:
We probably want to store the resolved template as well, so we can quickly determine if we need to rerun the templater.
```yaml
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: dynamic-catalog
  namespace: olm
  annotations:
    olm.catalogImageTemplate: "quay.io/kube-release-v{kube_major_version}/catalog:v{kube_major_version}.{kube_minor_version}"
...
status:
  olm.catalogImageTemplate:
    messages:
    resolvedImage: "quay.io/sample{kube_major_version}/catalog:{group:foo.example.com,version:v1,kind:Sample,name:MySample,namespace:ns,jsonpath:spec.foo.bar}"  
    resolvedCatalogImageTemplate: "quay.io/kube-release-v{kube_major_version}/catalog:v{kube_major_version}.{kube_minor_version}"  
```

So the CatalogSource Switcher Controller has these flows:

**New Annotated CatalogSource**
- CatalogSource Event
  - Has `metadata.annotations["olm.catalogImageTemplate"]`?  Yes
    -  Is valid?  Yes
    - Any kube* variables?   Yes
      - Start poller to watch the API Server for version changes (implementation TBD)
    - Any GVK variables?  Yes
      - Start watch for these specific resources.
    -  Has changed? (Resolved template <> Annotation template?)  Yes (it's new)
      -  Update image reference
      -  Update status
        - Store reconciled status

**Updated Annotated CatalogSource**
- CatalogSource Event
  - Has `metadata.annotations["olm.catalogImageTemplate"]`?  Yes
    -  Is valid?  Yes
    -  Has changed? (Resolved template <> Annotation template?)  Yes.  
      -  Update image reference
      -  Update status
        - Store reconciled status

**Deleted Annotated CatalogSource**
- CatalogSource Event
  - Has `olmmetadata.annotations["olm.catalogImageTemplate"]`?  No.
  - Has `status["olm.catalogImageTemplate"].resolvedCatalogImageTemplate?  Yes.  Deleted
    -  Remove resolvedCatalogImageTemplate from status
    -  Update status message to indicate that the template is no longer considered.
    -  Stop poller and watcher (store state in a configmap)

**Updated Kubernetes Version**
_Opportunity for caching and optimization here_
- Notification of version change:  1.18.1 to 1.18.2
  - Iterate through each CatalogSource and re-process each template.

**Updated Template Target Resource**
_Opportunity for caching and optimization here_
- Notification of resource update.  (Watch Event)
  - Iterate through each CatalogSource and re-process each template.

**Delete Template Target Resource**
_Opportunity for caching and optimization here_
- Notification of resource update.  (Watch Event)
  - Iterate through each CatalogSource and re-process each template.
    - Target resource is now gone, so the template can no longer be updated.

**Start controller**
- Check state configmap (pollers, watchers)
- Reestablish polling and watchers.
- Iterate through all CatalogSources to verify no lost updates/deletions
- Verify the kube version
- Verify all template target resource state.

Reply:
There's two conditions where we get an opportunity to evaluate the templates:

1. During sync

   My assumption is that we would be creating a new controller that runs in the catalog operator. When that controller is setup, we'd pass in the sync duration (using whatever value we decide on) and then every time the sync process kicks in we can evaluate the templates. See the [catalog operator CRD controller setup](https://github.com/operator-framework/operator-lifecycle-manager/blob/13549cb0a88e24667b0960b879286ec8c8bb4432/pkg/controller/operators/catalog/operator.go#L364) for a conceptual example. Note that all of the controllers use the same sync period (which is configurable) but that does not mean we have to use the exact same period (or even allow configuration of this period).

1. Event processing:

   We'd also be able to evaluate the templates whenever a watched resource changes.